src/OrchardCore.Modules/OrchardCore.Rewrite/Startup.cs

```diff
@@ -0,0 +1,60 @@
+using Microsoft.AspNetCore.Builder;
+using Microsoft.AspNetCore.Rewrite;
+using Microsoft.AspNetCore.Routing;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Options;
+using OrchardCore.Admin;
+using OrchardCore.Navigation;
+using OrchardCore.Security.Permissions;
+using OrchardCore.Settings;
+using OrchardCore.Rewrite.Rules;
+using OrchardCore.Modules;
+using OrchardCore.Rewrite.Models;
+using OrchardCore.Rewrite.Drivers;
+using OrchardCore.DisplayManagement.Handlers;
+
+namespace OrchardCore.Rewrite;
+
+public sealed class Startup : StartupBase
+{
+    public override int Order
+        => OrchardCoreConstants.ConfigureOrder.InfrastructureService;
+
+    private readonly AdminOptions _adminOptions;
+
+    public Startup(IOptions<AdminOptions> adminOptions)
+    {
+        _adminOptions = adminOptions.Value;
+    }
+
+    public override void ConfigureServices(IServiceCollection services)
+    {
+        services.AddScoped<INavigationProvider, AdminMenu>();
+        services.AddSiteDisplayDriver<RewriteSettingsDisplayDriver>();
+        services.AddScoped<IPermissionProvider, Permissions>();
+    }
+
+    public override async ValueTask ConfigureAsync(IApplicationBuilder app, IEndpointRouteBuilder routes, IServiceProvider serviceProvider)
+    {
+        var siteService = app.ApplicationServices.GetRequiredService<ISiteService>();
+        var modRewriteSettings = await siteService.GetSettingsAsync<RewriteSettings>();
+
+        var rewriteSettings = modRewriteSettings?.ApacheModRewrite;
+        if (rewriteSettings == null)
+        {
+            return;
+        }
+
+        using var apacheModRewrite = new StringReader(rewriteSettings);
+
+        var options = new RewriteOptions()
+            .AddApacheModRewrite(apacheModRewrite);
+
+        if (options.Rules.Count > 0)
+        {
+            // Exclude admin ui requests to prevent accidental access bricking by provided rules
+            options.Rules.Insert(0, new ExcludeAdminUIRule(_adminOptions.AdminUrlPrefix));
+            app.UseRewriter(options);
+        }
```

Comment:
I think this logic should be moved into a configuration class instead.

In other words, create a new sealed class that implement `IConfigureOptions<RewriteOptions>` then resolve `IOptions<RewriteOptions>` to add it to the pipeline.

Then, you can remove `ExcludeAdminUIRule` class since that logic would be in the configuration class.

Reply:
I tried that right now and that may not work. Looks like `IConfigureOptions/IAsyncConfigureOptions` `Configure/ConfigureAsync` kick in before document store is configured so I am not able to load configuration using: `await _siteService.GetSettingsAsync<UrlRewritingSettings>();` . Is there any walkaround for that?

Reply:
The configuration class will be invoked when a service resolves `IOptions<RewriteOptions>` for the very first time. Don't use `IAsyncConfigureOptions`. Use `IConfigureOptions` and get the settings from the document to populate.